### PR TITLE
[Pod] Finish MVP Integration

### PR DESCRIPTION
**Pull request by nicholashusin**
This is a collective PR that includes the work that the pod did together via pair programming.

The pod has decided to open this considerably large PR rather than separating it for now. This is because these changes are integrated and we feel that they will be harder to make sense of in smaller chunks.

If you think smaller chunks of PR will make it easier to review, please feel free to let the pod know and we will split it up.

Changes made:
- Remove unused artifacts in the codebase.
- Add integration for problem submissions. Users can now submit their solutions for problems and it will be stored in Datastore should it pass all test cases.
- Add integration for viewing solutions. Users can now view other people's solutions ranked by their complexity.
- Add access control for problem submissions. Users must be logged in to be able to submit a solution and can only submit it once.
- Add access control for viewing solutions. Users must have solved a certain problem to be able to see other people's solution for that problem.

Screenshots:
- [Viewing problems and solutions list](https://screenshot.googleplex.com/J6mHCc0eahM.png)
- [Viewing solutions for solved problems](https://screenshot.googleplex.com/RHs7ebLR1a0.png)

### CODE REVIEW

**Review by lalaAdasia00 — APPROVED** [review]

**Review by ifeomi — APPROVED** [review]

File: routes/problems_id_solutions.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  /**
+   * GET handler that displays problem dynamically according
+   * to the problem id given in the request parameter.
+   **/
+  app.get('/problems/:id/solutions', async function(request, response) {
+    const solutions = await getSolutions(request.params.id, DEFAULT_LIMIT, DEFAULT_SORT, DEFAULT_IS_DESCENDING);
+    response.render('solutions', {solutions: solutions});
+  });
+
+  app.post('/problems/:id/solutions', async function(request, response) {
```
Comment (chronologos) [inline routes/problems_id_solutions.js:52]: should be a get
https://google.aip.dev/131

File: routes/problems_id_solutions.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  return (await datastore.runQuery(query))[0];
+}
+
+module.exports = function(app) {
+  /**
+   * GET handler that displays problem dynamically according
+   * to the problem id given in the request parameter.
+   **/
+  app.get('/problems/:id/solutions', async function(request, response) {
+    const solutions = await getSolutions(request.params.id, DEFAULT_LIMIT, DEFAULT_SORT, DEFAULT_IS_DESCENDING);
```
Comment (chronologos) [inline routes/problems_id_solutions.js]: should be called `listSolutions`

File: routes/problems_id_solutions.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    .order(sortBy, {
+      descending: isDescending,
+    })
+    .limit(parseInt(limit))
+    .select(['email', 'complexity']);
+
+  return (await datastore.runQuery(query))[0];
+}
+
+async function getSubmittedCode(problemId, email) {
```
Comment (chronologos) [inline routes/problems_id_solutions.js:32]: for non-mvp, getSolutions and getSubmittedCode or their callers need to handle errors.

**Review by chronologos — APPROVED** [review]
lgtm with minor comments

File: routes/problems_id_solutions.js (hunk 1 above)
Comment (nicholashusin) [reply routes/problems_id_solutions.js:52]: Thanks for the feedback!

How should we go about implementing access control if we use a GET request for this part?
We have been sending authentication token via request body but I've found online that:
```
As per RFC 7231

GET is the primary mechanism of information retrieval and the focus of almost all performance optimizations.

A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.
```
This led me to believe that the only way to transmit authentication token with GET request would be via request parameter, which would be unsafe.

File: routes/problems_id_solutions.js (hunk 3 above)
Comment (nicholashusin) [reply routes/problems_id_solutions.js:32]: Noted, thanks!
I'll submit this issue in our Buganizer.

File: routes/problems_id_solutions.js (hunk 2 above)
Comment (nicholashusin) [reply routes/problems_id_solutions.js]: Fixed, thanks!

**Pull request merged**